File: web/src/io/devices/uart16550.ts
import { defaultReadValue } from "../ipc/io_protocol.ts";
import type { PortIoHandler } from "../bus/portio.ts";

export interface SerialOutputSink {
  write(port: number, data: Uint8Array): void;
}

export interface UartConfig {
  basePort: number;
  irq: number;
}

export const UART_COM1: UartConfig = { basePort: 0x3f8, irq: 4 };
export const UART_COM2: UartConfig = { basePort: 0x2f8, irq: 3 };
export const UART_COM3: UartConfig = { basePort: 0x3e8, irq: 4 };
export const UART_COM4: UartConfig = { basePort: 0x2e8, irq: 3 };

/**
 * Minimal 16550 UART model (COM ports).
 *
 * This is intended for BIOS/bootloader serial logging and basic debugging.
 * It models the core register set and emits TX bytes via an injected sink.
 */
export class Uart16550 implements PortIoHandler {
  readonly #cfg: UartConfig;
  readonly #sink?: SerialOutputSink;

  #dll = 0x00;
  #dlm = 0x00;
  #ier = 0x00;
  #fcr = 0x00;
  #lcr = 0x00;
  #mcr = 0x00;
  #scr = 0x00;

  #rxQueue: number[] = [];

  constructor(cfg: UartConfig, sink?: SerialOutputSink) {
    this.#cfg = cfg;
    this.#sink = sink;
  }

  get basePort(): number {
    return this.#cfg.basePort;
  }

  injectRx(byte: number): void {
    this.#rxQueue.push(byte & 0xff);
  }

  portRead(port: number, size: number): number {
    const p = port & 0xffff;
    switch (size) {
      case 1:
        return this.#readU8(p);
      case 2:
        return (this.#readU8(p) | (this.#readU8((p + 1) & 0xffff) << 8)) >>> 0;
      case 4:
        return (
          this.#readU8(p) |
          (this.#readU8((p + 1) & 0xffff) << 8) |
          (this.#readU8((p + 2) & 0xffff) << 16) |
          (this.#readU8((p + 3) & 0xffff) << 24)
        ) >>> 0;
      default:
        return defaultReadValue(size);
    }
  }

  portWrite(port: number, size: number, value: number): void {
    const p = port & 0xffff;
    const v = value >>> 0;
    switch (size) {
      case 1:
        this.#writeU8(p, v & 0xff);
        break;
      case 2:
        this.#writeU8(p, v & 0xff);
        this.#writeU8((p + 1) & 0xffff, (v >>> 8) & 0xff);
        break;
      case 4:
        this.#writeU8(p, v & 0xff);
        this.#writeU8((p + 1) & 0xffff, (v >>> 8) & 0xff);
        this.#writeU8((p + 2) & 0xffff, (v >>> 16) & 0xff);
        this.#writeU8((p + 3) & 0xffff, (v >>> 24) & 0xff);
        break;
      default:
        break;
    }
  }

  #dlab(): boolean {
    return (this.#lcr & 0x80) !== 0;
  }

  #readU8(port: number): number {
    const offset = (port - this.#cfg.basePort) & 0xffff;
    switch (offset) {
      case 0: // RBR / DLL
        if (this.#dlab()) return this.#dll;
        return (this.#rxQueue.shift() ?? 0) & 0xff;
      case 1: // IER / DLM
        if (this.#dlab()) return this.#dlm;
        return this.#ier;
      case 2: {
        // IIR (we do not currently model interrupts; report "no interrupt pending")
        const fifoEnabled = (this.#fcr & 0x01) !== 0;
        const fifoBits = fifoEnabled ? 0xc0 : 0x00;
        return (fifoBits | 0x01) & 0xff;
      }
      case 3:
        return this.#lcr & 0xff;
      case 4:
        return this.#mcr & 0xff;
      case 5: {
        // LSR: bit0=DR, bit5=THRE, bit6=TEMT
        let lsr = 0x60;
        if (this.#rxQueue.length > 0) lsr |= 0x01;
        return lsr;
      }
      case 6:
        return 0x00;
      case 7:
        return this.#scr & 0xff;
      default:
        return 0x00;
    }
  }

  #writeU8(port: number, value: number): void {
    const offset = (port - this.#cfg.basePort) & 0xffff;
    const v = value & 0xff;

    switch (offset) {
      case 0: // THR / DLL
        if (this.#dlab()) {
          this.#dll = v;
        } else {
          this.#sink?.write(this.#cfg.basePort, Uint8Array.of(v));
        }
        break;
      case 1: // IER / DLM
        if (this.#dlab()) this.#dlm = v;
        else this.#ier = v;
        break;
      case 2: // FCR
        this.#fcr = v;
        if ((v & 0x02) !== 0) this.#rxQueue = [];
        break;
      case 3:
        this.#lcr = v;
        break;
      case 4:
        this.#mcr = v;
        break;
      case 7:
        this.#scr = v;
        break;
      default:
        break;
    }
  }
}

File: crates/emulator/src/io/serial/uart16550.rs
use std::cell::{Cell, RefCell};
use std::collections::VecDeque;

use crate::io::PortIO;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct UartConfig {
    pub base_port: u16,
    pub irq: u8,
}

impl UartConfig {
    pub const COM1: Self = Self {
        base_port: 0x3F8,
        irq: 4,
    };
    pub const COM2: Self = Self {
        base_port: 0x2F8,
        irq: 3,
    };
    pub const COM3: Self = Self {
        base_port: 0x3E8,
        irq: 4,
    };
    pub const COM4: Self = Self {
        base_port: 0x2E8,
        irq: 3,
    };
}

pub struct Uart16550 {
    config: UartConfig,
    dll: Cell<u8>,
    dlm: Cell<u8>,
    ier: Cell<u8>,
    fcr: Cell<u8>,
    lcr: Cell<u8>,
    mcr: Cell<u8>,
    scr: Cell<u8>,
    rx_fifo: RefCell<VecDeque<u8>>,
    tx_callback: RefCell<Option<Box<dyn FnMut(u16, &[u8])>>>,
}

impl std::fmt::Debug for Uart16550 {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Uart16550")
            .field("config", &self.config)
            .field("dll", &self.dll.get())
            .field("dlm", &self.dlm.get())
            .field("ier", &self.ier.get())
            .field("fcr", &self.fcr.get())
            .field("lcr", &self.lcr.get())
            .field("mcr", &self.mcr.get())
            .field("scr", &self.scr.get())
            .field("rx_fifo_len", &self.rx_fifo.borrow().len())
            .finish_non_exhaustive()
    }
}

impl Uart16550 {
    pub fn new(config: UartConfig) -> Self {
        Self {
            config,
            dll: Cell::new(0),
            dlm: Cell::new(0),
            ier: Cell::new(0),
            fcr: Cell::new(0),
            lcr: Cell::new(0),
            mcr: Cell::new(0),
            scr: Cell::new(0),
            rx_fifo: RefCell::new(VecDeque::new()),
            tx_callback: RefCell::new(None),
        }
    }

    pub fn config(&self) -> UartConfig {
        self.config
    }

    pub fn set_tx_callback(&self, cb: impl FnMut(u16, &[u8]) + 'static) {
        *self.tx_callback.borrow_mut() = Some(Box::new(cb));
    }

    pub fn inject_rx(&self, byte: u8) {
        self.rx_fifo.borrow_mut().push_back(byte);
    }

    fn dlab(&self) -> bool {
        (self.lcr.get() & 0x80) != 0
    }

    fn read_u8(&self, port: u16) -> u8 {
        let offset = port.wrapping_sub(self.config.base_port);
        match offset {
            0 => {
                if self.dlab() {
                    self.dll.get()
                } else {
                    self.rx_fifo.borrow_mut().pop_front().unwrap_or(0)
                }
            }
            1 => {
                if self.dlab() {
                    self.dlm.get()
                } else {
                    self.ier.get()
                }
            }
            2 => {
                let fifo_enabled = (self.fcr.get() & 0x01) != 0;
                let fifo_bits = if fifo_enabled { 0xC0 } else { 0x00 };
                fifo_bits | 0x01
            }
            3 => self.lcr.get(),
            4 => self.mcr.get(),
            5 => {
                let mut lsr = 0x60;
                if !self.rx_fifo.borrow().is_empty() {
                    lsr |= 0x01;
                }
                lsr
            }
            6 => 0,
            7 => self.scr.get(),
            _ => 0,
        }
    }

    fn write_u8(&mut self, port: u16, value: u8) {
        let offset = port.wrapping_sub(self.config.base_port);
        match offset {
            0 => {
                if self.dlab() {
                    self.dll.set(value);
                } else if let Some(cb) = self.tx_callback.borrow_mut().as_mut() {
                    cb(self.config.base_port, &[value]);
                }
            }
            1 => {
                if self.dlab() {
                    self.dlm.set(value);
                } else {
                    self.ier.set(value);
                }
            }
            2 => {
                self.fcr.set(value);
                if (value & 0x02) != 0 {
                    self.rx_fifo.borrow_mut().clear();
                }
            }
            3 => self.lcr.set(value),
            4 => self.mcr.set(value),
            7 => self.scr.set(value),
            _ => {}
        }
    }
}

impl PortIO for Uart16550 {
    fn port_read(&self, port: u16, size: usize) -> u32 {
        match size {
            1 => self.read_u8(port) as u32,
            2 => {
                let lo = self.read_u8(port) as u32;
                let hi = self.read_u8(port.wrapping_add(1)) as u32;
                lo | (hi << 8)
            }
            4 => {
                let b0 = self.read_u8(port) as u32;
                let b1 = self.read_u8(port.wrapping_add(1)) as u32;
                let b2 = self.read_u8(port.wrapping_add(2)) as u32;
                let b3 = self.read_u8(port.wrapping_add(3)) as u32;
                b0 | (b1 << 8) | (b2 << 16) | (b3 << 24)
            }
            _ => 0,
        }
    }

    fn port_write(&mut self, port: u16, size: usize, val: u32) {
        match size {
            1 => self.write_u8(port, val as u8),
            2 => {
                self.write_u8(port, val as u8);
                self.write_u8(port.wrapping_add(1), (val >> 8) as u8);
            }
            4 => {
                self.write_u8(port, val as u8);
                self.write_u8(port.wrapping_add(1), (val >> 8) as u8);
                self.write_u8(port.wrapping_add(2), (val >> 16) as u8);
                self.write_u8(port.wrapping_add(3), (val >> 24) as u8);
            }
            _ => {}
        }
    }
}
